src/services/CatalogService.ts

```diff
@@ -1,30 +1,38 @@
+import * as Effect from "@effect/io/Effect";
 import { CatalogProcessError, ErrorCode } from "../model/domain/errors.js";
 import { convertToClientEServiceSeed } from "../model/domain/models.js";
 import { ApiEServiceSeed } from "../model/types.js";
 import { eserviceSeedToCreateEvent } from "../repositories/adapters/adapters.js";
+import { DB } from "../repositories/db.js";
 import { eventRepository } from "../repositories/events.js";
 import { readModelGateway } from "./ReadModelGateway.js";
 
 export const catalogService = {
-  async createEService(apiEservicesSeed: ApiEServiceSeed): Promise<void> {
-    const organizationId = await readModelGateway.getOrganizationID();
-
-    const eserviceSeed = convertToClientEServiceSeed(
-      apiEservicesSeed,
-      organizationId
-    );
-
-    const eservice = await readModelGateway.getEServiceByName(
-      eserviceSeed.name
-    );
-
-    if (eservice !== undefined) {
-      throw new CatalogProcessError(
-        `Error during EService creation with name ${eserviceSeed.name}`,
-        ErrorCode.DuplicateEserviceName
+  createEService(
+    apiEservicesSeed: ApiEServiceSeed
+  ): Effect.Effect<DB, CatalogProcessError, void> {
+    return Effect.gen(function* (_) {
+      const organizationId = yield* _(readModelGateway.getOrganizationID());
+      const eserviceSeed = convertToClientEServiceSeed(
+        apiEservicesSeed,
+        organizationId
       );
-    }
-
-    return eventRepository.createEvent(eserviceSeedToCreateEvent(eserviceSeed));
+      const eservice = yield* _(
+        readModelGateway.getEServiceByName(eserviceSeed.name)
+      );
+      return _(
+        Effect.noneOrFailWith(
+          eservice,
+          () =>
+            new CatalogProcessError(
+              `Error during EService creation with name ${eserviceSeed.name}`,
+              ErrorCode.DuplicateEserviceName
+            )
+        ),
+        Effect.flatMap(() =>
+          eventRepository.createEvent(eserviceSeedToCreateEvent(eserviceSeed))
```

Comment:
qui mi sto perdendo qualcosa: `eventRepository.createEvent` non ha come errore `unknown`?  come finiamo su `CatalogProcessError`? immagino che gli errori del repository andrebbero ignorati come unexpected con `die` ma non lo trovo

Reply:
Urca, hai ragione, questo dovrebbe beccarlo TypeScript. Forse è figlio di un refactor, domani investigo

Reply:
Oddio, ho capito cosa è successo.
Stavo ritornando con

```ts
return _(...);
```

invece che 

```ts
return yield* _(...);
```

e nessuno si lamentava perché temo che `void` si "mangiasse" il tipo di ritorno (cioè probabilmente stavamo tornando `Effect<DB, unknown, void` al posto di `void` e nessuno si lamentava perché void è un po' speciale in TypeScript, vedi https://github.com/microsoft/TypeScript/issues/8584)

Aggiungendo `yield*` torna tutto a posto e in effetti abbiamo un errore (che ho gestito nell'ultimo commit).

Un po' una footgun questa 🤔, anche se poi l'effetto non avrebbe girato e quindi in un test l'avremmo beccato.